forge-game/src/main/java/forge/game/ability/effects/RepeatEachEffect.java

```diff
@@ -139,6 +137,36 @@ else if (sa.hasParam("DefinedCards")) {
             }
         }
 
+        if (sa.hasParam("RepeatTypes")) {
+            final String repTypes = sa.getParam("RepeatTypes");
+            final List<String> validTypes = new ArrayList<>();
+            if (repTypes.startsWith("ThisTurnCast")) {
+                final String[] workingCopy = repTypes.split("_");
+                final String validFilter = workingCopy[1];
+                List<Card> res = CardUtil.getThisTurnCast(validFilter, source, sa);
+                for (final Card c : res) {
+                    for (String t : c.getType()) {
+                        validTypes.add(t);
+                    }
+                }
+            } else {
+                System.err.println("Warning: Code needed in RepeatEachEffect for " + source);
+            }
+
+            String storedType = source.getChosenType();
+            while (validTypes.size() >= 2) {
+                String chosenT = player.getController().chooseSomeType("card", sa, validTypes, null);
+                source.setChosenType(chosenT);
+                AbilityUtils.resolve(repeat);
+                validTypes.remove(chosenT);
+            }
+            if (validTypes.size() == 1) {
```

Comment:
That looks ugly: chooseSomeType should already be able to shortcut out of prompting when there's only a single option left